power_grid_model_c/power_grid_model/include/power_grid_model/main_model_impl.hpp

```diff
@@ -1499,4 +1306,22 @@ class MainModelImpl<ExtraRetrievableTypes<ExtraRetrievableType...>, ComponentLis
     }
 };
 
+template <calculation_input_type CalcInputType>
+static auto calculate_param(auto const& c, auto const&... extra_args)
+    requires requires {
+        { c.calc_param(extra_args...) };
+    }
+{
+    return c.calc_param(extra_args...);
+}
+
+template <calculation_input_type CalcInputType>
+static auto calculate_param(auto const& c, auto const&... extra_args)
+    requires requires {
+        { c.template calc_param<typename CalcInputType::sym>(extra_args...) };
+    }
+{
+    return c.template calc_param<typename CalcInputType::sym>(extra_args...);
+}
```

Comment:
Moved them to `detail` as suggested. 

On the other hand, I had not noticed what you point out and, after a bit of reading, I found that this only works because the `calculate_param` functions are templated. Why? Because the compiler first parses the whole file, but only instantiates them once the template type is known later on. So as long as the functions are templated and within the same translation unit where they are called, there is no need to forward declare them. Pretty cool find :).

Edit: Funny enough, if I add it to another namespace, i.e. detail, it doesn't work anymore. Hence I just moved it all the way up.

Reply:
welp 🤦